Infrastructure-Scripts/Wait-AzureDevOpsDeployment.ps1

```diff
@@ -14,52 +21,89 @@
+$Url = "$AzureDevOpsOrganisationUri$AzureDevOpsProjectName/_apis/distributedtask/environments/$EnvironmentId/environmentdeploymentrecords?top=500&api-version=6.1-preview.1"
 $RetryCounter = 0
 
-while ($RetryCounter -lt 30){
+if ($PatToken) {
+    $Base64PatToken = [Convert]::ToBase64String([Text.Encoding]::ASCII.GetBytes(":$($PatToken)"))
+    $Headers = @{ Authorization = "Basic $Base64PatToken" }
+}
+else {
+    $Headers = @{ Authorization = "Bearer $env:SYSTEM_ACCESSTOKEN" }
+}
+
+while ($RetryCounter -lt $RetryLimit) {
     Write-Verbose "Attempt $RetryCounter"
+    $EnvironmentDeployments = @()
     try{
         #Invoke call to Azure DevOps Rest API to get all deployment data for given environment.
-        $EnvironmentDeployments = Invoke-RestMethod -Method GET -Uri $Url -Headers  @{ Authorization = "Bearer $env:SYSTEM_ACCESSTOKEN" } -TimeoutSec 30
+        Write-Verbose "Requesting: $Url"
+        $EnvironmentDeployments += (Invoke-RestMethod -Method GET -Uri $Url -Headers $Headers -TimeoutSec 30 -ResponseHeadersVariable ResponseHeaders).value
+        while ($ResponseHeaders.'x-ms-continuationtoken') {
+            $ContinationUrl = "$Url&continuationToken=$($ResponseHeaders.'x-ms-continuationtoken')"
+            $EnvironmentDeployments += (Invoke-RestMethod -Method GET -Uri $ContinationUrl -Headers $Headers -TimeoutSec 30 -ResponseHeadersVariable ResponseHeaders).value
+            if ($ResponseHeaders.'Retry-After') {
+                Write-Warning "Request is being rate limited"
+                ##TO DO: implement logic to handle rate limiting responses - https://docs.microsoft.com/en-us/azure/devops/integrate/concepts/rate-limits?view=azure-devops#api-client-experience
+            }
+        }
     }
     catch{
         Write-Error "Response code $($_.Exception.Response.StatusCode.Value__) received. Terminating process, deployment will continue."
         Write-Output $_
         break;
     }
-    #Filter down results of API call to just contain relevant pipeline runs with matching Pipeline names and only ones that are still running.
-    $RunningEnvironmentDeployments = $EnvironmentDeployments.value | Where-Object {$_.definition.name -eq $PipelineName -and -not $_.result}
+
+    #Filter down results of API call to just contain relevant pipeline runs with matching Pipeline names and only ones that are still running.  This does not include queued jobs.
+    $RunningEnvironmentDeployments = $EnvironmentDeployments | Where-Object {$_.definition.name -eq $PipelineName -and -not $_.result}
+    Write-Host "Running jobs:`n$($RunningEnvironmentDeployments | Select-Object -Property startTime, @{ label="pipelineName"; expression={$_.definition.name}}, result, id, stageName, jobName, @{label="ownerName";expression={$_.owner.name}}, @{label="ownerId";expression={$_.owner.id}}, planId | Where-Object { $_.pipelineName -eq $PipelineName } | Format-Table | Out-String)"
```

Comment:
Is the logging of every job in the pipeline run required for logging abit overkill? Maybe its best if just the pipeline runs itself and not its individual jobs?

Reply:
I've included all the jobs because the problem we're trying to solve is to ensure that `Pipeline Run 2` doesn't start deploying to `Stage A` until `Pipeline Run 1` has finished all of it's `Jobs` for `Stage A`

Though maybe that info only needs to be available in verbose logs

Reply:
I think it could be handy to see the table of which jobs are in particular are running but understand why you'd say it's better kept to verbose logs